File: backend/internal/handlers/health.go
package handlers

import (
	"context"
	"database/sql"
	"net/http"
	"time"

	"github.com/gin-gonic/gin"
)

type HealthHandler struct {
	db *sql.DB
}

func NewHealthHandler(db *sql.DB) *HealthHandler {
	return &HealthHandler{db: db}
}

type HealthResponse struct {
	Status    string                 `json:"status"`
	Timestamp time.Time              `json:"timestamp"`
	Version   string                 `json:"version"`
	Database  DatabaseHealth         `json:"database"`
	Services  map[string]interface{} `json:"services"`
}

type DatabaseHealth struct {
	Connected bool   `json:"connected"`
	Latency   string `json:"latency,omitempty"`
	Error     string `json:"error,omitempty"`
}

// GetSystemHealth returns the health status of the system
// GET /api/v1/health
func (h *HealthHandler) GetSystemHealth(c *gin.Context) {
	startTime := time.Now()

	// Check database connection
	dbHealth := h.checkDatabaseConnection()

	// Determine overall status
	status := "healthy"
	statusCode := http.StatusOK
	if !dbHealth.Connected {
		status = "unhealthy"
		statusCode = http.StatusServiceUnavailable
	}

	response := HealthResponse{
		Status:    status,
		Timestamp: time.Now(),
		Version:   "1.0.0",
		Database:  dbHealth,
		Services: map[string]interface{}{
			"api": map[string]interface{}{
				"status":      "operational",
				"uptime":      time.Since(startTime).String(),
				"environment": "production",
			},
		},
	}

	c.JSON(statusCode, response)
}

// checkDatabaseConnection verifies database connectivity
func (h *HealthHandler) checkDatabaseConnection() DatabaseHealth {
	// Handle nil database
	if h.db == nil {
		return DatabaseHealth{
			Connected: false,
			Error:     "database connection is nil",
		}
	}

	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
	defer cancel()

	startTime := time.Now()

	// Ping database
	if err := h.db.PingContext(ctx); err != nil {
		return DatabaseHealth{
			Connected: false,
			Error:     err.Error(),
		}
	}

	latency := time.Since(startTime)

	return DatabaseHealth{
		Connected: true,
		Latency:   latency.String(),
	}
}

File: backend/src/handlers/health.ts
import type { Context } from 'hono';
import { pool } from '../db/connection.js';

export async function getSystemHealth(c: Context) {
  const startTime = Date.now();

  // Check database connection
  let dbHealth: { connected: boolean; latency?: string; error?: string };
  try {
    const dbStart = Date.now();
    const client = await pool.connect();
    try {
      await client.query('SELECT 1');
      const latency = Date.now() - dbStart;
      dbHealth = { connected: true, latency: `${latency}ms` };
    } finally {
      client.release();
    }
  } catch (err) {
    dbHealth = { connected: false, error: (err as Error).message };
  }

  const status = dbHealth.connected ? 'healthy' : 'unhealthy';
  const statusCode = dbHealth.connected ? 200 : 503;

  const response: Record<string, unknown> = {
    status,
    timestamp: new Date().toISOString(),
    version: '1.0.0',
    database: dbHealth,
    services: {
      api: {
        status: 'operational',
        uptime: `${Date.now() - startTime}ms`,
        environment: process.env.NODE_ENV || 'production',
      },
    },
  };

  return c.json(response, statusCode as 200);
}
